crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3298 @@
+                        let t = self - &ex_rhs;
+                        t.is_zero()
+                    }
+                }
+            }
+            (_, _) => false,
+        }
+    }
+}
+
+impl PartialEq<f64> for SymbolExpr {
+    fn eq(&self, r: &f64) -> bool {
+        match self.eval(true) {
+            Some(v) => v == *r,
+            None => false,
+        }
+    }
+}
+
+impl PartialEq<Complex64> for SymbolExpr {
+    fn eq(&self, r: &Complex64) -> bool {
+        match self.eval(true) {
+            Some(v) => v == *r,
+            None => false,
+        }
+    }
+}
+
+// comparison rules for sorting equation
+impl PartialOrd for SymbolExpr {
+    fn partial_cmp(&self, rhs: &Self) -> Option<Ordering> {
+        match self {
+            SymbolExpr::Value(l) => match rhs {
+                SymbolExpr::Value(r) => l.partial_cmp(r),
+                _ => Some(Ordering::Less),
+            },
+            SymbolExpr::Symbol(l) => match rhs {
+                SymbolExpr::Value(_) => Some(Ordering::Greater),
+                SymbolExpr::Symbol(r) => l.partial_cmp(r),
+                SymbolExpr::Unary { op: _, expr } => self.partial_cmp(expr),
+                _ => Some(Ordering::Less),
+            },
+            SymbolExpr::Unary { op: _, expr } => match rhs {
+                SymbolExpr::Value(_) => Some(Ordering::Greater),
+                SymbolExpr::Unary { op: _, expr: rexpr } => expr.partial_cmp(rexpr),
+                _ => (expr.as_ref()).partial_cmp(rhs),
+            },
+            SymbolExpr::Binary {
+                op,
+                lhs: ll,
+                rhs: lr,
+            } => match rhs {
+                SymbolExpr::Value(_) | SymbolExpr::Symbol(_) => match op {
+                    BinaryOp::Mul | BinaryOp::Div | BinaryOp::Pow => Some(Ordering::Greater),
+                    _ => Some(Ordering::Equal),
+                },
+                SymbolExpr::Unary { op: _, expr } => self.partial_cmp(expr),
+                SymbolExpr::Binary {
+                    op: _,
+                    lhs: rl,
+                    rhs: rr,
+                } => {
+                    let ls = match ll.as_ref() {
+                        SymbolExpr::Value(_) => lr.to_string(),
+                        _ => self.to_string(),
+                    };
+                    let rs = match rl.as_ref() {
+                        SymbolExpr::Value(_) => rr.to_string(),
+                        _ => rhs.to_string(),
+                    };
+                    if rs > ls && rs.len() > ls.len() {
+                        Some(Ordering::Less)
+                    } else if rs < ls && rs.len() < ls.len() {
+                        Some(Ordering::Greater)
+                    } else {
+                        Some(Ordering::Equal)
+                    }
+                }
+            },
+        }
+    }
+}
+
+impl From<&str> for SymbolExpr {
+    fn from(v: &str) -> Self {
+        SymbolExpr::Symbol(Box::new(v.to_string()))
+    }
+}
```

Comment:
The existence of this is further evidence to me that `ParameterExpression` should _not_ implement `From<&str>` by parsing a string. This one is a more reasonable behavior IMO.

Reply:
I agree